sdk/daml-script/test/src/main/scala/com/digitalasset/daml/lf/engine/script/test/UpgradesMatrixIT.scala

```diff
@@ -0,0 +1,288 @@
+      // GrpcLedgerClient doesn't support Name refs, so we replace with pkgId and use the explicitPackageId flag instead
+      commands = apiCommands.toList.map {
+        case n @ ApiCommand.Create(`tplRef`, _) =>
+          ScriptLedgerClient.CommandWithMeta(
+            n.copy(templateRef = testHelper.v1TplId.toRef),
+            false,
+          )
+        case n @ ApiCommand.Exercise(`tplRef`, _, _, _) =>
+          ScriptLedgerClient.CommandWithMeta(
+            n.copy(typeRef = testHelper.v1TplId.toRef),
+            false,
+          )
+        case n @ ApiCommand.ExerciseByKey(`tplRef`, _, _, _) =>
+          ScriptLedgerClient.CommandWithMeta(
+            n.copy(templateRef = testHelper.v1TplId.toRef),
+            false,
+          )
+        case n @ ApiCommand.CreateAndExercise(`tplRef`, _, _, _) =>
+          ScriptLedgerClient.CommandWithMeta(
+            n.copy(templateRef = testHelper.v1TplId.toRef),
+            false,
+          )
+        case n => ScriptLedgerClient.CommandWithMeta(n, true)
+      }
+      result <- scriptClient.submit(
+        actAs = OneAnd(setupData.alice, Set()),
+        readAs = Set(),
+        disclosures = disclosures,
+        optPackagePreference =
+          Some(List(cases.commonDefsPkgId, cases.templateDefsV2PkgId, cases.clientPkgId)),
+        commands = commands,
+        prefetchContractKeys = List(),
+        optLocation = None,
+        languageVersionLookup =
+          _ => Right(LanguageVersion.defaultOrLatestStable(LanguageMajorVersion.V2)),
+        errorBehaviour = ScriptLedgerClient.SubmissionErrorBehaviour.Try,
+      )
+    } yield result
+
+  override def assertResultMatchesExpectedOutcome(
+      result: Either[
+        ScriptLedgerClient.SubmitFailure,
+        (Seq[ScriptLedgerClient.CommandResult], ScriptLedgerClient.TransactionTree),
+      ],
+      expectedOutcome: ExpectedOutcome,
+  ): Assertion = {
+    expectedOutcome match {
+      case ExpectSuccess =>
+        result shouldBe a[Right[_, _]]
+      case ExpectUpgradeError =>
+        inside(result) { case Left(ScriptLedgerClient.SubmitFailure(_, error)) =>
+          error should (
+            be(a[SubmitError.UpgradeError.ValidationFailed]) or
+              be(a[SubmitError.UpgradeError.DowngradeDropDefinedField]) or
+              be(a[SubmitError.UpgradeError.DowngradeFailed])
+          )
+        }
+      case ExpectPreprocessingError =>
+        inside(result) { case Left(_) =>
+          // error shouldBe a[EE.Preprocessing] // I dont know what we mean by preprocessing here
+          Succeeded
+        }
+      case ExpectPreconditionViolated =>
+        inside(result) { case Left(ScriptLedgerClient.SubmitFailure(_, error)) =>
+          error shouldBe a[SubmitError.TemplatePreconditionViolated]
+        }
+      case ExpectUnhandledException =>
+        inside(result) { case Left(ScriptLedgerClient.SubmitFailure(_, error)) =>
+          error shouldBe a[SubmitError.FailureStatusError]
```

Comment:
Had to change this from an UnhandledException to a FailureStatusError due to the new Grpc error work, but would like a second pair of eyes to be sure I haven't missed anything obvious.

Reply:
Paul reviewed and approved